File: sdk/typescript/src/telemetry.ts
export interface TelemetryClientConfig {
    endpoint?: string;
    apiKey: string;
}

export interface TelemetryEvent {
    id: string;
    definition_id: string;
    name: string;
    event_type: string;
    selector?: string;
    url_pattern?: string;
    visual_guide?: string;
}

export interface TelemetryDefinition {
    id: string;
    account_id: string;
    experiment_id: string;
    description: string;
    is_active: boolean;
    events: TelemetryEvent[];
    created_at: string;
    updated_at: string;
}

export interface TelemetryDefinitionsResponse {
    definitions: TelemetryDefinition[];
}

export class BeakerTelemetry {
    private endpoint: string;
    private apiKey: string;

    constructor(config: TelemetryClientConfig) {
        this.endpoint = config.endpoint ?? '/api/sdk/telemetry';
        this.apiKey = config.apiKey;
    }

    async fetchDefinitions(experimentId: string): Promise<TelemetryDefinitionsResponse> {
        const url = `${this.endpoint}?experiment_id=${encodeURIComponent(experimentId)}`;
        const response = await fetch(url, {
            method: 'GET',
            headers: {
                'x-beaker-key': this.apiKey,
            },
        });

        if (!response.ok) {
            const error = await response.text();
            throw new Error(`Telemetry request failed: ${error}`);
        }

        return response.json();
    }

    async getActiveDefinitions(experimentId: string): Promise<TelemetryDefinition[]> {
        const result = await this.fetchDefinitions(experimentId);
        return result.definitions ?? [];
    }
}

File: sdk/python/beaker/telemetry.py
import requests
from typing import Optional, Dict, Any, List
import logging

logger = logging.getLogger(__name__)


class TelemetryClientConfig:
    def __init__(
        self,
        api_key: str,
        endpoint: Optional[str] = None
    ):
        self.api_key = api_key
        self.endpoint = endpoint or "/api/sdk/telemetry"


# TelemetryEvent shape:
# {
#   "id": str,
#   "definition_id": str,
#   "name": str,
#   "event_type": str,
#   "selector": Optional[str],
#   "url_pattern": Optional[str],
#   "visual_guide": Optional[str],
# }

# TelemetryDefinition shape:
# {
#   "id": str,
#   "account_id": str,
#   "experiment_id": str,
#   "description": str,
#   "is_active": bool,
#   "events": List[TelemetryEvent],
#   "created_at": str,
#   "updated_at": str,
# }


class BeakerTelemetry:
    def __init__(self, config: TelemetryClientConfig):
        self.config = config
        self.endpoint = config.endpoint

    def fetch_definitions(self, experiment_id: str) -> Dict[str, Any]:
        headers = {
            "x-beaker-key": self.config.api_key
        }

        try:
            response = requests.get(
                self.endpoint,
                params={"experiment_id": experiment_id},
                headers=headers,
                timeout=10
            )
            response.raise_for_status()
            return response.json()
        except Exception as e:
            logger.error(f"Failed to fetch telemetry definitions: {e}")
            raise e

    def get_active_definitions(self, experiment_id: str) -> List[Dict[str, Any]]:
        try:
            result = self.fetch_definitions(experiment_id)
            return result.get("definitions", [])
        except Exception:
            return []
